File: leetcode-submissions/112_Path_Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(!root){
            return false;
        }
        queue<pair<TreeNode*,int>>q;
        q.push({root,root->val});
        while(not q.empty()){
            auto [currNode,currSum]=q.front();
            q.pop();
            if(currSum==targetSum){
                return true ;
            }
            if(currNode->left){
                q.push({currNode->left,currSum+currNode->left->val});
            }
            if(currNode->right){
                q.push({currNode->right,currSum+currNode->right->val});
            }
        }
        return false ;
    }
};

File: leetcode-submissions/112_Path_Sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {

    if(!root) return false;
    const queue:[(TreeNode | null),number][] =[[root,root.val]];
    while(queue.length){
        let currSize:number = queue.length;
        while(currSize--){
            const [currNode,currSum]=queue.shift()!;
            if(currSum===targetSum) return true ;
            if(currNode.left){
                queue.push([currNode.left,currSum+currNode.left.val]);
            }
            if(currNode.right){
                queue.push([currNode.right,currSum+currNode.right.val]);
            }
        }
    }
    return false;

};